### ISSUE TITLE
[RFC] Support MiniMax H3 FL2VA with DiffusionNFT and FlowGRPO

### ISSUE THREAD

**Issue by NancyFyong** [verl-project/verl-omni#396]
## Summary

Add first-frame, last-frame, and first+last-frame conditioning for the MiniMax H3
FL2VA checkpoint under two training algorithms:

1. **DiffusionNFT** trains from the final clean generated video/audio target rows.
   Reference-frame condition rows remain fixed and are excluded from forward
   noising and loss.
2. **FlowGRPO** records separate stochastic video/audio reverse-SDE transitions.
   Reference-frame condition rows remain fixed and are excluded from trajectories
   and policy log-probabilities.

Both algorithms share the same official vLLM-Omni FL2VA request contract, packed
sequence semantics, token-ID-native multimodal prompt path, FSDP2 LoRA Actor,
and Actor-to-rollout weight synchronization.

## Motivation

MiniMax H3 jointly generates video and stereo audio, while its FL2VA checkpoint
also conditions generation on one or two keyframes. This is not ordinary I2V
conditioning: vLLM-Omni packs condition-video rows, generated-video rows, audio
rows, text embeddings, multimodal token tags, three-dimensional positions, and
per-row timesteps into one DiT sequence.

A correct RL integration must preserve all of those semantics during Actor
replay. Treating FL2VA as T2VA with extra images is incorrect because it can:

- update condition rows that should remain pinned to reference anchors;
- apply ordinary forward noise to reference rows;
- assign data timesteps instead of conditioning timesteps;
- lose vision-token modality tags or keyframe temporal positions;
- include condition rows in policy trajectories and log-probabilities;
- decode and re-tokenize prompts, changing multimodal presentation IDs;
- replay a different video/audio sigma schedule from rollout.

Supporting both DiffusionNFT and FlowGRPO also validates that one FL2VA transport
contract can serve a direct-preference objective and a policy-gradient objective
without conflating their different training targets.

## Relationship to existing work and non-duplication

- [RFC #347](https://github.com/verl-project/verl-omni/issues/347) proposes
  MiniMax H3 **T2VA FlowGRPO** on GPU and NPU and explicitly lists first/last-frame
  conditioning as a non-goal. This RFC is GPU-first and specifically owns the
  FL2VA conditioning/replay contract for both DiffusionNFT and FlowGRPO.
- Draft [PR #368](https://github.com/verl-project/verl-omni/pull/368) implements
  H3 T2VA FlowGRPO with joint audio-video CPS and CLAP/ImageBind rewards. The
  FlowGRPO part of this RFC should reuse whichever T2VA base is accepted; it
  should not submit another copy of the T2VA infrastructure. Its material delta
  is fixed FL2VA condition rows, official keyframe layouts, separate modality SDE
  replay, and target-only densities.
- Draft [PR #383](https://github.com/verl-project/verl-omni/pull/383) implements
  H3 T2VA DiffusionNFT. The NFT part of this RFC is a direct extension of that
  branch from unconditional six-field target latents to explicit FL2VA
  conditioning state and target-only noising/loss.
- [RFC #391](https://github.com/verl-project/verl-omni/issues/391) proposes a
  stable vLLM-Omni diffusion-RL interface. This implementation currently uses a
  custom H3 pipeline because the required typed trajectory and custom-payload
  seams are not yet available upstream. It should migrate when that interface
  lands rather than preserving a permanent forward-loop fork.

### Goals

- MiniMax H3 FL2VA on GPU with LoRA and FSDP2.
- First-frame conditioning: `frame_indices=[0]`.
- Last-frame conditioning: `frame_indices=[-1]`.
- First+last conditioning: `frame_indices=[0, -1]`.
- Online DiffusionNFT with default and old-policy LoRA adapters.
- FlowGRPO with separate video/audio SDE legs and rollout importance correction.
- Token-ID-native multimodal prompts.
- Official vLLM-Omni condition packing, keyframe positions, and condition-noise
  augmentation.
- PickScore over sampled video frames as the baseline reward.
- Exact base-checkpoint and LoRA transport validation between Diffusers and
  vLLM-Omni.
- Data converters and launch recipes for all three keyframe modes.

### Non-goals

- Ref2VA or reference-video conditioning.
- NPU support in the first version.
- Full-parameter Actor training.
- Training the text encoder or video/audio VAEs.
- More than one output per request; the current H3 pipeline supports one request
  at a time and the adapter rejects `num_outputs_per_prompt != 1`.
- Online FP8 rollout for the custom pipeline; BF16 is the validated baseline.
- Audio-quality or audio-video synchronization rewards. PickScore evaluates only
  prompt-video alignment.
- Claiming media quality from 2–4-step smoke runs. This checkpoint is not a
  low-step distilled model and requires 50 inference steps for quality runs.

### Training flow

```text
prompt + one/two keyframe images
  -> H3 token-ID-native agent loop
  -> vLLM-Omni FL2VA image encoding and official packed layout
  -> DiffusionNFT final-state payload OR FlowGRPO reverse-SDE trajectory
  -> PickScore reward
  -> FSDP2 Diffusers Actor replay
  -> rank-64 LoRA update
  -> tensor payload synchronization
  -> TP=4 vLLM-Omni fused H3 LoRA loading
```

### Input and prompt contract

The dataset stores the textual prompt separately from structured image inputs.
The H3 agent loop tokenizes user text exactly once, with no special tokens, and
passes `prompt_token_ids` plus multimodal image data to rollout. It must not
round-trip IDs through `tokenizer.decode()`.

vLLM-Omni then constructs the FL2VA presentation using:

- `<Picture i>: ` text prefixes;
- image-pad vision slots;
- `VIDEO(0)` modality tags for visual presentation tokens;
- `TEXT(1)` tags for textual tokens;
- keyframe-dependent temporal positions.

The model's processor metadata is incomplete in the released checkpoint. An H3
processor preparation hook supplies `model_type=qwen3_vl` when needed so
`Qwen3VLProcessor` can be created before `RLHFDataset` processes images.

### Official packed condition layout

The implementation consumes vLLM-Omni's official input matrix instead of
reimplementing request interpretation. For FL2VA, packed media rows are ordered
as:

```text
condition-video rows -> audio rows -> generated-video rows
```

Only generated-video rows have `update_mask=true`. `-1` resolves to the final
latent frame. The first/last temporal positions and multimodal token tags are
transported to Actor replay without approximation.

At each denoise step:

- condition rows are restored to their keyframe anchors;
- condition rows use `max(current video data-fraction timestep, 0.999)`;
- the request's visual-condition noise augmentation is applied to both the
  anchor values and conditioning timestep semantics;
- only generated-video rows receive scheduler updates.

### Schedule contract

Video and audio use independent shifted schedules:

- video flow shift: `12.0`;
- audio flow shift: `3.0`.

The rollout payload transports the exact shifted video/audio sigma arrays.
Actor replay installs those arrays instead of reconstructing a nominal schedule.
If a future checkpoint supplies `base_schedule`, checkpoint metadata may override
`num_inference_steps`; rollout and Actor must use the same resulting schedule.

### Checkpoint and LoRA layouts

Rollout and Actor intentionally use separate transformer directories:

- `MODEL_PATH/transformer`: official vLLM-Omni fused-QKV/GEGLU layout;
- `ACTOR_TRANSFORMER_PATH`: converted Diffusers
  `MiniMaxH3Transformer3DModel` layout.

The base policy is equivalent after:

- grouped-QKV reordering;
- GEGLU gate/up half swapping;
- parameter-name conversion;
- `rope.inv_freq` synthesis.

LoRA synchronization maps Diffusers targets into vLLM-Omni fused modules:

- `to_q`, `to_k`, `to_v` -> TP-local fused `qkv_proj` slices;
- `ff.net.0.proj` gate/up halves -> fused `fc1` slices;
- `to_out.0` -> row-parallel `out_proj`;
- `ff.net.2` -> row-parallel `fc2`.

The validated baseline uses rank 64, alpha 128, and default/old adapters where
required. Diffusers modules listed in `_keep_in_fp32_modules` retain checkpoint
FP32 precision, and FSDP mixed precision must not cast those islands to BF16.

### Rollout payload

The NFT custom pipeline emits the final generated clean state and everything
needed to reproduce FL2VA Actor inputs:

- `latents_clean` and `train_timesteps`;
- latent shape metadata;
- prompt embeddings, masks, and modality tags;
- condition-video rows and keyframe frame indices;
- exact video/audio schedule metadata.

### Actor replay

The Actor reconstructs condition rows ahead of the generated targets. It applies
forward noising only to generated video/audio target rows. Condition rows remain
at their augmented anchors and receive the conditioning timestep. The model may
predict velocities for every video row, but NFT crops condition-row outputs
before loss calculation.

This preserves the direct-preference objective: reference images define fixed
conditioning state, while only generated media is optimized.

### Rollout trajectory

The H3 rollout adapter runs two stochastic legs:

- a video SDE over generated-video rows only;
- an audio SDE over all generated-audio rows.

Condition-video rows never enter either stochastic state. They are reinserted
before every DiT call and pinned after every scheduler update.

The rollout payload includes:

- packed target-only latent trajectories;
- video and audio timesteps;
- weighted transition log-probabilities;
- condition rows, keyframe indices, and multimodal prompt state;
- exact video and audio sigma schedules;
- target video/audio row counts for deterministic unpacking.

Per-modality log-probabilities are mean-reduced by their schedulers and combined
using configurable video/audio weights. This avoids weighting one modality only
because it has more latent elements.

### Actor replay and rollout correction

Actor replay reconstructs the full packed model input from each target-only
trajectory state, inserts the pinned condition rows, recomputes Diffusers
velocities, crops generated-video outputs, and evaluates both modality
transition densities against the exact rollout schedules.

The fused vLLM-Omni rollout and Diffusers Actor are close but not bit-identical.
Therefore the recipe records rollout log-probabilities and enables sequence-level
rollout importance correction with `bypass_mode=False`; the Actor recomputes the
old-policy log-probabilities instead of treating rollout values as exact Actor
values.

## Configuration and resource baseline

Proposed upstream baseline:

| Item | Value |
| --- | --- |
| vLLM-Omni pin | `728640f4d9bb7c1b68d646e2a4c59ce1ce45de9c` |
| Actor | Diffusers + FSDP2, BF16 with declared FP32 islands |
| LoRA | rank 64, alpha 128 |
| Rollout TP | 4 |
| GPUs | 8 x 96 GB for the default one-step training recipe |
| CPU offload | enabled |
| Actor attention | `_flash_3_varlen_hub` (FA3) |
| Rollout attention | `FLASH_ATTN_3_HUB` (FA3) |
| Rollout batch | two prompts x `ROLLOUT_N=4`, divisible by 8 Actor ranks |
| Max sequences | 1 |
| Frames | request 96 at 24 FPS; vLLM-Omni aligns to the H3 `17n+5` boundary |
| Inference steps | 50 for quality; 2–4 only for contract smoke tests |
| Reward | PickScore |

TP=2 is not the default because the colocated dummy-load phase can place a full
text encoder beside one DiT shard before CPU offload activates, leaving too
little memory for Actor-to-rollout synchronization. TP=4 has passed the complete
rank-64 synchronization path.

Training and rollout deliberately use the matched FA3 pair. The Diffusers Actor
loads `_flash_3_varlen_hub`, while vLLM-Omni loads `FLASH_ATTN_3_HUB`; both are
backed by `kernels-community/flash-attn3`. The exact `kernels` dependency and Hub
kernel artifact must be installed or pre-cached on every node before an offline
run. `native` Actor plus `TORCH_SDPA` rollout remains an explicit debugging and
portability fallback, not the proposed production default. Mixed pairs such as
`native` plus `FLASH_ATTN_3_HUB` are unsupported because Actor replay and rollout
must use corresponding attention implementations.

### Static and CPU validation

- NFT broad CPU suite: **679 passed, 2 skipped**.
- FlowGRPO broad CPU suite: **643 passed, 2 skipped**.
- Focused latest-source suites: **63 NFT tests** and **77 FlowGRPO tests** passed.
- Upstream CPU-safe MiniMax H3 suites: **148 tests** passed.
- Changed-file pre-commit checks passed on both branches, except the repository's
  pre-existing stale `autogen-trainer-cfg` baseline, which was intentionally not
  rewritten by this work.

### Checkpoint and synchronization validation

- Full base-checkpoint parity matched all **535 vLLM tensors** against all **638
  Diffusers tensors**, covering **33,122,992,912 values** with exact shape,
  dtype, and value equality after layout conversion.
- Real-GPU rank-64/alpha-128 Actor-to-rollout LoRA synchronization passed at
  **TP=4 on all four ranks**, including QKV, FC1, output projection, FC2,
  TP-local slicing, scaling, and row-parallel reduction.
- Real TP=2 full-checkpoint rollout passed first `[0]`, last `[-1]`, and
  first+last `[0, -1]` modes with finite clean latents, active image/text tags,
  and the expected condition-row counts.

### End-to-end validation

- **8-GPU FlowGRPO:** completed generation, PickScore reward, old-policy
  recomputation, rollout correction, Actor update, and rollout weight update at
  global step 1 with TP=4 and `ROLLOUT_N=4`.
- **8-GPU DiffusionNFT:** completed generation, PickScore reward, Actor update,
  and generation dumping at global step 1 with TP=4 and `ROLLOUT_N=4`.
- A second **4-GPU DiffusionNFT** one-step run passed using the explicit
  `native`/`TORCH_SDPA` fallback.
- The completed training-loop runs above validate the fallback attention pair,
  not the proposed matched FA3 default. A full Actor+rollout run with
  `_flash_3_varlen_hub`/`FLASH_ATTN_3_HUB` is required before the RFC is ready
  to merge.
- Those complete training-loop runs used the immediately preceding reviewed
  vLLM-Omni pin. After advancing to `728640f4`, focused CPU suites, all three
  real full-checkpoint FL2VA modes, exact checkpoint parity, and TP=4 LoRA
  synchronization were rerun; a complete one-step training loop has not yet
  been repeated on that final pin.
- Separate full-checkpoint **50-step** first-frame FL2VA quality generation
  produced a decoded 448x256 video and stereo 32 kHz audio. The earlier 2-step
  training outputs are contract smokes only and are not quality evidence.

## Acceptance criteria

1. All three keyframe modes match upstream vLLM-Omni packing exactly.
2. Prompt IDs are never decoded and re-tokenized between dataset and rollout.
3. Condition rows remain equal to their augmented anchors at every reverse step.
4. NFT forward noising and loss exclude condition rows.
5. FlowGRPO trajectories and log-probabilities exclude condition rows.
6. Actor replay consumes rollout-emitted video/audio sigma schedules exactly.
7. Base-checkpoint parity consumes every tensor on both sides.
8. Rank-64 LoRA synchronization passes at the recipe's TP=4 setting.
9. A real full-checkpoint rollout-to-reward-to-Actor-update run passes for each
   algorithm with Actor `_flash_3_varlen_hub` and rollout
   `FLASH_ATTN_3_HUB`.
10. FA3 rollout/Actor replay reports finite log-probabilities and an importance
    ratio within the configured rollout-correction clipping range.
11. Quality examples use 50 inference steps unless a checkpoint-provided
    distilled schedule explicitly says otherwise.

## Upstreaming plan

To avoid duplicating #368 and #383, do not merge the two experimental branches
as independent monolithic PRs.

1. **Land/rebase the H3 T2VA base.** Reuse #383 for NFT and coordinate the
   FlowGRPO base with #368.
2. **Extract shared FL2VA contracts.** Submit token-ID-native H3 agent-loop,
   processor preparation, task/frame-index config, official keyframe metadata,
   common packing tests, dtype preservation, and latest vLLM-Omni compatibility
   as one reviewable layer where practical.
3. **Stack the NFT extension.** Add final-state conditioning payload and
   target-only forward noising/loss on top of #383.
4. **Stack the FlowGRPO extension.** Add pinned condition reconstruction,
   target-only dual-SDE trajectories, exact schedule replay, and rollout
   correction on top of the accepted H3 FlowGRPO base.
5. **Add recipes and special E2E tests.** Keep full-checkpoint parity and TP=4
   LoRA tests as opt-in GPU/special-E2E validation; keep contract tests CPU-safe.
6. **Migrate away from loop overrides.** Adopt the stable rollout interface from
   #391 once vLLM-Omni exposes the required scheduler, typed payload, and
   pre-tokenized multimodal seams.

## Risks and mitigations

| Risk | Mitigation |
| --- | --- |
| Upstream vLLM-Omni contract drift | Pin a reviewed commit and run upstream H3 contract suites before each bump |
| Silent official-vs-Diffusers weight mismatch | Streaming exact full-checkpoint parity test |
| Condition rows accidentally trained | Explicit update mask plus target-only trajectory/loss tests |
| Prompt semantic drift | Token-ID-native agent loop and decode-rejection tests |
| Rollout/Actor schedule mismatch | Transport and install exact video/audio sigma arrays |
| Rollout/Actor numerical mismatch | Old-policy Actor recomputation and rollout importance correction |
| FP32 projection islands recast by FSDP | Preserve `_keep_in_fp32_modules` and disable FSDP param casting for those models |
| TP=2 memory pressure | TP=4 plus CPU offload as the documented baseline |
| FA3 Hub artifact unavailable offline | Pin `kernels`, prefetch `kernels-community/flash-attn3` into a shared cache on every node, and fail before Ray startup if it is missing |
| FA3 rollout/Actor numerical drift | Use the matched FA3 pair, recompute old-policy log-probabilities, and enforce rollout-correction metrics in the GPU smoke |

## Contribution

I can split the validated fork implementation into the shared, NFT, and
FlowGRPO layers above, keep the special-E2E parity/synchronization tests, and
provide exact commands and results in each PR. Before submission, the human
submitter will review every changed line and repeat the duplicate-work checks.

AI assistance (Pi Coding Agent) was used to implement, test, and draft this RFC.

**Comment by NancyFyong** [comment]
Submission-scope clarification: the common diffusion Actor `use_remove_padding` configuration, capability checks, engine hooks and model-independent attention-boundary utility are being submitted separately on a main-based branch. This is not another H3 FL2VA or packed-forward implementation: the model-specific work stays in #592, which can later rebase onto the shared interface. The interface PR has no dependency on #592, carries none of its adapter/rollout changes, and does not close this RFC. Request-level rollout batching such as #562 is also outside its scope.
